llvm/lib/MC/WasmObjectWriter.cpp

```diff
@@ -1482,10 +1482,6 @@ uint64_t WasmObjectWriter::writeOneObject(MCAssembler &Asm,
     LLVM_DEBUG(dbgs() << "Processing Section " << SectionName << "  group "
                       << Section.getGroup() << "\n";);
 
-    // .init_array sections are handled specially elsewhere.
-    if (SectionName.starts_with(".init_array"))
-      continue;
```

Comment:
Can would write this data if (and only if) there are symbols that point to it?   Wasm is very sensitive to binary size and even a few bytes of wasted data section is not desirable, especially since unused data is impossible to DCE after linking.

Reply:
Yes, I think we can do that. Good idea! 

I've added code to include the `.init_array` sections' data only if there are symbols referencing that section.

It also means we don't have to tweak the existing tests, which is nice.